[Marcus Chua] iP

marcuschj opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](http://github.com))

DukePro frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](http://github.com)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😀 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

marcuschj committed:
add second hello

marcuschj committed:
Level-1

marcuschj committed:
Level 1 customisation

marcuschj committed:
Level-2

marcuschj committed:
Level 3

marcuschj committed:
Level 4

marcuschj committed:
A-TextUiTesting

marcuschj committed:
Level 5

marcuschj committed:
Level 6

marcuschj committed:
A-Enums

marcuschj committed:
Level-7

marcuschj committed:
Level-8

marcuschj committed:
Merge branch 'branch-Level-7' into master

marcuschj committed:
Merge branch 'branch-Level-8' into master

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/Task.java

marcuschj committed:
Update test cases in input.txt

marcuschj committed:
A-MoreOOP

marcuschj committed:
Revert "A-MoreOOP"

This reverts commit 9f3a2859a102f8bafbaf2af5e3dc7def280a14ba.

marcuschj committed:
Add more OOP to Ui

marcuschj committed:
Add packages

marcuschj committed:
Add junit test

marcuschj committed:
Add pre-release for A-Jar

marcuschj committed:
Add JavaDoc for classes

marcuschj committed:
Add JavaDoc to methods

marcuschj committed:
Enforce coding standards

marcuschj committed:
Add find feature

marcuschj committed:
Merge branch 'branch-A-JavaDoc' into master

marcuschj committed:
Merge branch 'branch-A-CodingStandard' into master

# Conflicts:
#	src/main/java/duke/DeadlineTest.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

marcuschj committed:
Merge branch 'branch-Level-9' into master

# Conflicts:
#	src/main/java/duke/Ui.java

marcuschj committed:
Update input.txt

marcuschj committed:
Add package for test classes

marcuschj committed:
Update runtest.bat with correct file paths

marcuschj committed:
Merge remote-tracking branch 'origin/add-gradle-support' into master

marcuschj committed:
Add Gradle to iP

marcuschj committed:
Merge branch 'branch-A-Gradle' into master

marcuschj committed:
Performed CheckStyle

marcuschj committed:
Merge branch 'branch-A-CheckStyle' into master

marcuschj committed:
Add GUI as Level-10

marcuschj committed:
Merge branch 'branch-Level-10' into master

marcuschj committed:
Add UI improvements for JavaFX GUI

marcuschj committed:
Add tag for A-Varargs but no applicable usage of varargs

ziqing26 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * A class of exceptions that are thrown for invalid user inputs when parsed.
+ */
+public class DukeException extends Exception {
```
Just curious any reason why you declare multiple exception classes in one file instead of separating them into different files?

ziqing26 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,224 @@
+                    }
+                } else {
+                    throw new DukeException.DukeTaskFailException();
+                }
```
Perhaps you could make it more readable by reducing some nested `if else` statements?

ziqing26 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,224 @@
+     *
+     * @param message String command that is received from the Scanner in Ui.
+     */
+    public String handleInput(String message) {
```
Just a suggestion: Perhaps you could abstract out some parts into other functions as this method currently is a bit lengthy.

ziqing26 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * Loads the existing <code>Task</code>s from the <code>TaskList</code> when Duke is ran.
+     *
+     * @throws FileNotFoundException if <code>filePath</code> specified is invalid.
```
What does this function return?

ziqing26 commented on src/test/java/DeadlineTest.java:
```diff
@@ -0,0 +1,17 @@
+public class DeadlineTest {
+
+    @Test
+    public void testDeadlineToString() {
```
Maybe give it a better name to include what is the input and what is the expected behaviour?

ziqing26 reviewed (commented):
In general, I think you could make more abstractions:)

marcuschj committed:
Assertions: add assertions using assert feature

Current code will not terminate the code execution if a possible bug is discovered. This could be unsafe.

Assertions are used to define assumptions about the program state so that the runtime can verify them.

Let's
* do some refactoring to abstract variables as suitable checkpoints for assertions
* add assertions in the classes Duke, Parser, Storage and TaskList.

Refer to this resource on the importance of assertions and how to include them:
https://nus-cs2103-ay2122s1.github.io/website/schedule/week5/topics.html#w5-6-assertions

marcuschj committed:
Improve code quality: improve readability and reducing unsafe practices

Current code has many similar lines and has poor abstractions.
In particular, handleInput() function in Parser class has almost 200 lines.

Code is difficult to debug and cannot be understood by collaborators.

Requires refactoring and restructuring, which can break code into bite-size chunks for anyone reading the code.

Let's
* Critically examine the code
* Refactor to improve the code quality where necessary, focusing on extracting methods from handleInput

kjpvi commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    @Override
+    public void start(Stage stage) {
```
I like the way you organize the start function - very intuitive

marcuschj committed:
printList and showSearchResults: add streams

Current printList and showSearchResults are cumbersome and involve verbose for loops over the ArrayList of Tasks.

Using streams to map each index in the list to the correct String format makes for cleaner code and increases code quality.

Streams are also used to show understanding of new Java 8 features.

Let's
* Rewrite the existing printList and showSearchResults in the Storage and Ui classes respectively
* Test the implementation via GUI

marcuschj committed:
Merge pull request #1 from marcuschj/branch-A-Assertions

Assertions: add assertions using assert feature

marcuschj committed:
Merge branch 'master' of https://github.com/marcuschj/ip into branch-A-Streams

# Conflicts:
#	src/main/java/duke/TaskList.java

marcuschj committed:
Merge branch 'master' into branch-A-CodeQuality

marcuschj committed:
Merge pull request #2 from marcuschj/branch-A-CodeQuality

Improve code quality: improve readability and reducing unsafe practices

marcuschj committed:
Merge branch 'master' of https://github.com/marcuschj/ip into master

marcuschj committed:
Merge branch 'master' into branch-A-Streams

marcuschj committed:
Merge pull request #3 from marcuschj/branch-A-Streams

printList and showSearchResults: add streams

marcuschj committed:
Fix checkstyle errors from A-Streams, A-CodeQuality and A-Assertions

marcuschj committed:
Fix GUI bugs

marcuschj committed:
Add C-Tags

Tagging of tasks is now enabled. Users can add hashtags to tasks, and search for tasks based on the hashtag too.

marcuschj committed:
Merge branch 'branch-BCD-Extension' into master

timiditi commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,113 @@
+     * Loads the existing <code>Task</code>s from the <code>TaskList</code> when Duke is ran.
+     *
+     * @throws FileNotFoundException if <code>filePath</code> specified is invalid.
+     */
```
Don't forget your @return statement in your java docs!

timiditi commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        String tagsList = this.tags
+            .stream()
```
Take note that the module standard is to leave 8 spaces for wrapped lines

timiditi commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,174 @@
+
+    /**
+     * Prints on user interface the output for Welcome Message 1.
+     */
```
Do you want to add @return statements for your java docs in this class?

timiditi reviewed (commented):
LGTM! Just a couple of minor coding standard nitpicks.

marcuschj committed:
Update GUI to chat style

marcuschj committed:
Improve code quality in Duke and MainWindow classes

marcuschj committed:
Improve GUI

marcuschj committed:
Add product screenshot

marcuschj committed:
Edit template list

marcuschj committed:
Add task to template list

marcuschj committed:
Remove task from template list.txt

marcuschj committed:
Add stage title for GUI

marcuschj committed:
Add new jar file

marcuschj committed:
Set theme jekyll-theme-hacker

marcuschj committed:
Add UG

marcuschj committed:
Merge branch 'master' of https://github.com/marcuschj/ip into master

marcuschj committed:
Add Release
